File: leetcode/201-300/277-find-the-celebrity/bench/celebrity.py
#!/usr/bin/env python3
"""LeetCode 277 bench mirror — Python. Same algorithm; included for corpus
completeness rather than because the number is interesting."""

N = 2500000
INSTANCES = 64


def knows(star, a, b):
    if b == star:
        return True
    if a == star:
        return False
    return ((a * 1103515245 + b * 12345) % 2147483647) % 2 == 0


def find_celebrity(n, star):
    cand = 0
    for i in range(1, n):
        if knows(star, cand, i):
            cand = i
    for j in range(n):
        if j != cand:
            if knows(star, cand, j):
                return -1
            if not knows(star, j, cand):
                return -1
    return cand


def main():
    sink = 0
    for i in range(INSTANCES):
        star = (i * 7919) % N
        sink = (sink + (i * 1000003 + find_celebrity(N, star)) % 1000000007) % 1000000007
    print(sink)


main()

File: leetcode/201-300/277-find-the-celebrity/bench/celebrity.rs
// LeetCode 277 bench mirror — Rust. Same algorithm as celebrity.kara.
const N: i64 = 2500000;
const INSTANCES: i64 = 64;

fn knows(star: i64, a: i64, b: i64) -> bool {
    if b == star { return true; }
    if a == star { return false; }
    let h = (a * 1103515245i64 + b * 12345i64) % 2147483647i64;
    h % 2 == 0
}

fn find_celebrity(n: i64, star: i64) -> i64 {
    let mut cand = 0i64;
    for i in 1..n {
        if knows(star, cand, i) { cand = i; }
    }
    for j in 0..n {
        if j != cand {
            if knows(star, cand, j) { return -1; }
            if !knows(star, j, cand) { return -1; }
        }
    }
    cand
}

fn main() {
    let mut sink = 0i64;
    for i in 0..INSTANCES {
        let star = (i * 7919) % N;
        sink = (sink + (i * 1000003 + find_celebrity(N, star)) % 1000000007) % 1000000007;
    }
    println!("{}", sink);
}
